Fixes nested relative package module imports/Fixes incorrect module __package__ attribute

rkbennett opened the pull request:
This is a fix for #40, This fix essentially starts at the name of the module being imported and walks backwards looking in sys.modules where `sys.modules[path].__package__ == path`, and then uses that as the package since it would be the nearest level. This has an additional benefit of fixing nested relative package module imports as `__package__` is what python uses to determine where to import from when presented with a relative path.

rkbennett committed:
Fixes nested package module imports

operatorequals commented:
Hello @rkbennett !
Thank you so much for the PR!

I run the tests for your changes and some of them failed! Do not get bothered by the massive amount of Xs, they are the same unit tests that fail and they all test the archived module import (`zip`, `tar`, etc)!

Do you want to take a look?

rkbennett commented:
Yup, I'll check into these this evening!

rkbennett committed:
fixes issue with importing packages

Fixes issue with importing top level packages directly

rkbennett commented:
Got to it a bit earlier @operatorequals, I ended up running into it myself :D. So what was happening is that the code change assumes you're going to import true packages (`__init__.py`) which reference objects locally from nested packages. Obviously there is a use case where you're going to import something like a script.py directly  as the "package", for these instances I went ahead and put the code I added in an if statement that ensures the module size is greater than 1, if it is 1 or less, it reverts back to the old code which uses the first index (which would be the only index) in the module name.

rkbennett committed:
adds reference to nested package module

adds reference to mod2 within b

rkbennett committed:
add mod2 module under b

rkbennett committed:
changes value to uppercase B

rkbennett commented:
@operatorequals  also added a little bit to the tests so it implements the nested modules change, so under package b I added an additional module "mod2.py", and then made it only contain a mod2val var definition which is 'Module B', I then updated mod.py under package b to import it thusly `from .mod2 import mod2val` and changed the return value of the mod.py function to be `return mod2val`, this should return the same value it used to while using python relative package references.

rkbennett committed:
forces subpackage imports in test_package

add `from . import a, b` to force subpackage imports from test_package

rkbennett committed:
adds mod import to a package

adds mod import to a package

rkbennett committed:
adds mod import to a package

adds mod import to a package

operatorequals commented:
You are golden!
I'll review, merge and release in the following days!

rkbennett commented:
Interesting, so I was looking a bit more into that error, and amusingly enough, I had never used the httpimport.load() function because I could never get it to work for me. I looked into it a bit more and see that it only loads it into sys.modules. The issue that you run into, is that any modules that import functions from another source in the same package are going to break importing. Also noticed that `package_module/__init__.py` only has `__all__ = ['module1', 'module2']` but as far as I can understand the items in the array that are passed must exist in the script which is defining `__all__`, which would mean that `__all__ = ['module1','module2']` doesn't currently export anything. You can double check this by running `dir(sys.modules['test_package'])` to verify that test_module is not exporting module1 or module 2. However, if you drop a function inside `test_package/__init__.py` called module1, you'll notice that it actually does export a module1 function, which can then be called like this `sys.modules['test_package'].module1()`. That's a long round about way of saying that I don't think loading packages into sys.modules and not loading them into the base namespace will work on packages that do intra-package function imports. Sorry for the long-winded comment @operatorequals, thought I was going crazy for a minute. 🙃

rkbennett committed:
adds subpackage assertions for basic http, removes relative import from top level package

rkbennett commented:
@operatorequals, I went ahead and removed the relative import from test_package's `__init__` so the load test would still work, and I added a couple more unittest assertions under the simple http function to also test the relative path imports under test_package.b. Hopefully that works for both use cases and should pass unit testing 🤞

rkbennett commented:
I forgot I came across that issue in the past and fixed it locally, let me dig up where I did that fix and apply it to my repo :D

rkbennett committed:
squashes httpimport.py commits

rkbennett committed:
removes replace exclusion on final '.' in _mod_to_filename, also switches os.sep to '/' for url

rkbennett commented:
Okay, that one is submitted as well. I split the logic for getting filenames for modules/packages between zip and url, another issue I had ran into in the past was that, when you run this on Windows, os.sep messes with the url path sometimes. I also removed the exclusion of replacing the final . with a / when it's importing from url. @operatorequals

operatorequals commented:
It's soo close! Only the pre-0.9.0  check fails!
Maybe the reason for that is that `test_package.b` is not set in the first argument (which can be a list of importable packages).

rkbennett committed:
removes relative package import test from pre 9.0.0 due to the fact 9.0.0 can't support it

rkbennett commented:
Yeah, came to the realization that for pre-0.9.0 it wouldn't support the relative path imports anyway, since that's what this PR is adding to 0.9.2 😄,  so I just removed that check from pre 0.9.0

rkbennett commented:
Looks like all the tests ran, just a pypi error on 3.9

operatorequals commented:
All green!
What you are seeing looks like an issue of the workflow.

operatorequals commented:
Great!
Can you squash the commits to:
- one that has changes to the code
- one that has changes to unit tests
- one that has changes of the test package

So it is more tidy for the history. If you don't want I can squash and merge altogether - you've done so much already!

rkbennett commented:
Yup, I can take a look at it here in a bit.

rkbennett commented:
Hmm, looks like my squash didn't apply properly, would you mind doing a squash on import so I don't muck this up any more 😆

operatorequals commented:
> Hmm, looks like my squash didn't apply properly, would you mind doing a squash on import so I don't muck this up any more 😆 

aaaand it's gone

rkbennett commented:
Thanks, sorry for all the back and forth on this one heh

operatorequals commented:
A release has been made with your changes - `v0.9.3`
https://pypi.org/project/httpimport/0.9.3/

Also I added you in the contributors part in `README.md`.
Thank you so much for your time!
